text/0000-cargo-token-from-process.md

```diff
@@ -0,0 +1,142 @@
+- Feature Name: `cargo_token_from_process`
+- Start Date: 2019-07-22
+- RFC PR: [rust-lang/rfcs#2730](https://github.com/rust-lang/rfcs/pull/2730)
+- Cargo Issue: [rust-lang/cargo#0000](https://github.com/rust-lang/cargo/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a cargo setting to fetch registry authentication tokens by calling an
+external process.
+
+# Motivation
+[motivation]: #motivation
+
+Some interactions with a registry require an authentication token, and Cargo
+currently stores such token in plaintext in the [`.cargo/credentials`][creds]
+file. While Cargo properly sets permissions on that file to only allow the
+current user to read it, that's not enough to prevent other processes ran by
+the same user from reading the token.
+
+This RFC aims to provide a way to configure Cargo to instead fetch the token
+from any secrets storage system, for example a password manager or the system
+keyring.
+
+[creds]: https://doc.rust-lang.org/stable/cargo/reference/config.html#credentials
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Suppose a user has their authentication token stored in a password manager, and
+the password manager provides a command, `/usr/bin/cargo-creds`, to decrypt and
+print that token in a secure way. Instead of also storing the token in
+plaintext, the user can add this snippet to their own `.cargo/credentials` to
+authenticate with crates.io:
+
+```toml
+[registry]
+token-from-process = "/usr/bin/cargo-creds"
+```
+
+When authentication is required Cargo will execute the command and use its
+output as the token, which will never be stored by Cargo on disk. If the
+command requires arguments, for example `password-manager creds crates-io`, you
+can add them in a list:
+
+```toml
+[registry]
+token-from-process = ["password-manager", "creds", "crates-io"]
+```
+
+It will be possible to use `token-from-process` on both crates.io and
+alternative registries.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new key, `token-from-process`, will be added to the `[registry]` and
+`[registries.NAME]` sections of the `.cargo/credentials` configuration file.
+When a `token` key is also present, the latter will take precedence over
+`token-from-process` to maintain backward compatibility, and a warning will be
+issued to let the user know about that.
+
+The `token-from-process` key accepts either a string containing the binary to
+call or a list containing the binary name and the arguments to provide to it.
```

Comment:
FWIW Cargo's intepretation of this today is that a bare string is a whitespace-separated list of arguments and a list of strings is the exact list of arguments. We'll probably want to stay consistent with the rest of Cargo in that respect?

Reply:
My worry about splitting by whitespace is it just breaks in unexpected ways if people wants to use a shell:

```toml
token-from-process = 'bash -c "echo foo | head -n 1"'
```

With some CLI password managers this is going to be really common, and splitting `["bash", "-c", "\"echo", "foo", "|", "head" "-n", "1\""]` is just unintuitive. I'd prefer to either properly split quotes (by using crates like [shlex](https://docs.rs/shlex)) or just not split at all.

Reply:
I think that's possible, yes, but I think it's more important to either to stay consistent with Cargo or fully commit to some form of parsing.

Reply:
Personally I'd prefer to just commit to split like a POSIX shell. It's not that complex and there is already an implementation available in the [shlex crate](https://docs.rs/shlex).

<sub>Ugh, your reply was lost in the notifications... Sorry for the delay responding.</sub>

Reply:
I understand, yes, but i personally at least feel that using shlex is a local optimum which misses how nothing else in Rust uses shlex. For example `-C link-args`, `RUSTFLAGS`, etc. Nothing across our tooling which accepts multiple arguments has ever used shell escapes, and this seems like a pretty odd case to be the first one to use it.

Reply:
I would second the objection to `shlex` or any other parsing.

I'd actually prefer to prohibit "single string with whitespace splitting" and *only* allow a list of arguments, but if we don't want to do that, then let's be consistent with the rest of Cargo.

Reply:
@joshtriplett so, is the current text of the RFC what you'd like?